### ISSUE TITLE
eval STRING compilation failure breaks overloading

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10733]
Migrated from [rt.perl.org#78398](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=78398) (status was 'resolved')

Searchable as RT78398$

**Comment by p5pRT** [comment]
### From hdp@glaive.weftsoar.net

#### Created by hdp@glaive.weftsoar.net



Using eval STRING inside an overload method causes stack corruption when STRING
has compile\-time errors\.

Commenting out the eval makes it work; or changing it to just eval "require Foo&#8203;::Bar"
putting it inside \{ local $@&#8203;; eval \.\.\. \} makes it abort&#8203;:
perl&#8203;: pp\_ctl\.c&#8203;:2073&#8203;: Perl\_pp\_leaveloop&#8203;: Assertion \`\(\(\(cx\)\->cx\_u\.cx\_subst\.sbu\_type & 0xC\) == 0x4\)' fail\.
perl 5\.10\.1 and 5\.8\.9 segfault here instead of aborting\.
Any compile\-time error works\, like "BEGIN \{ \!@&#8203;\#\!@&#8203;\# \}" or "use Foo&#8203;::Bar"

use overload \(q\{""\} => 'str'\);
sub new \{ bless \{\} \}
sub str \{
&nbsp;    eval "BEGIN \{ require Foo&#8203;::Bar \}";
&nbsp;    return 1;
\}

main\->new\(\) \. "";

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.10.1:

Configured by hdp at Sun Apr 11 07:47:29 EDT 2010.

Summary of my perl5 (revision 5 version 10 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=2.6.31.5-linode21, archname=i686-linux
    uname='linux glaive 2.6.31.5-linode21 #1 smp mon oct 26 18:17:01 utc 2009 i686 gnulinux '
    config_args='-de -Dprefix=/home/hdp/perl5/perlbrew/perls/perl-5.10.1'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.9.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.9'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.10.1:
    /home/hdp/perl5/perlbrew/perls/perl-5.10.1/lib/5.10.1/i686-linux
    /home/hdp/perl5/perlbrew/perls/perl-5.10.1/lib/5.10.1
    /home/hdp/perl5/perlbrew/perls/perl-5.10.1/lib/site_perl/5.10.1/i686-linux
    /home/hdp/perl5/perlbrew/perls/perl-5.10.1/lib/site_perl/5.10.1
    .


Environment for perl 5.10.1:
    HOME=/home/hdp
    LANG=en_US.UTF8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/hdp/perl5/perlbrew/bin:/home/hdp/perl5/perlbrew/perls/current/bin:/home/hdp/bin:/home/hdp/local/bin:/home/hdp/local/sbin:/var/lib/gems/1.8/bin:/home/hdp/.inker/perl/bin:/home/hdp/bin:/home/hdp/local/bin:/home/hdp/local/sbin:/var/lib/gems/1.8/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    PERL_CPANM_OPT=--prompt --skip-installed --mirror http://cpan.cpantesters.org
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From hdp@pobox.com

I tested this with all of 5\.12\.2\, 5\.10\.1\, and 5\.8\.9\, and it breaks in
the same way under all of them\.

**Comment by p5pRT** [comment]
### From @rafl

For me\, it fails slightly differently\.

"Can't return outside a subroutine at" from the return in the overloaded
method\.

This is probably the same problem you're seeing\, but with a different
effect\. It might indicate that the exception is unwinding a little too far?

Also\, I can't produce this on blead\. I can't see any recent changes to
amagic\_call and friends that would fix anything like that\. Is it
possible that the general exception sanity fixes took care of this as well?

A bisect run\, anyone?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From hdp@pobox.com

On Sat Oct 16 02&#8203;:08&#8203;:25 2010\, rafl wrote&#8203;:
> "Can't return outside a subroutine at" from the return in the overloaded
> method\.

Sorry\, I forgot to say what the actual error is\. :\)

That's what I get too\.  The segfault/abort is only if I try to localize
$@&#8203; around the eval\.

**Comment by p5pRT** [comment]
### From @rafl

Apparently this is already fixed in 5\.13\.0\. Here's the bisect result\.

git bisect start
\# bad&#8203;: \[a4ef29484100fcd961129071b2ca69064250b195\] Add 5\.13\.0 to perlhist
git bisect bad a4ef29484100fcd961129071b2ca69064250b195
\# good&#8203;: \[6d52c880307229a35c23215c596700e716bd7c32\] Removing the RC
marker from patchlevel\.h
git bisect good 6d52c880307229a35c23215c596700e716bd7c32
\# bad&#8203;: \[d1515be42a17e7bb3fa584aea980f54524603f34\] mark two magic\.t tests
as TODO
git bisect bad d1515be42a17e7bb3fa584aea980f54524603f34
\# bad&#8203;: \[1bb125e2afe6197deaf55852a3f8a9c52736bfdc\] Note how to deal with
broken dbm\.h on OpenSUSE
git bisect bad 1bb125e2afe6197deaf55852a3f8a9c52736bfdc
\# bad&#8203;: \[11035fcf28d4d5fe35c7f6719dbd07b704a8f266\] remove 'enable taint
if modify gid/uid' feature
git bisect bad 11035fcf28d4d5fe35c7f6719dbd07b704a8f266
\# good&#8203;: \[099be4f1d597471eb719c9a344b7c1b55e11ba24\] PL\_defoutgv isn't
always a GV\.
git bisect good 099be4f1d597471eb719c9a344b7c1b55e11ba24
\# bad&#8203;: \[27e904532594b7fb224bdf9a05bf3b5336b8a39e\] fix RT 23810&#8203;: eval and
tied methods
git bisect bad 27e904532594b7fb224bdf9a05bf3b5336b8a39e
\# good&#8203;: \[91e35ba127b7082418836f7f9f428e4d2f9b5745\] more mods to \-Dl
debugging output
git bisect good 91e35ba127b7082418836f7f9f428e4d2f9b5745

27e904532594b7fb224bdf9a05bf3b5336b8a39e is the first bad commit
commit 27e904532594b7fb224bdf9a05bf3b5336b8a39e
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Thu Apr 8 13&#8203;:16&#8203;:56 2010 \+0100

&nbsp;    fix RT 23810&#8203;: eval and tied methods

&nbsp;    Something like the following ended up corrupted&#8203;:
&nbsp;        sub FETCH \{ eval 'BEGIN\{syntax err\}' \}
&nbsp;    The croak on error popped back the context stack etc to the EVAL
pushed by
&nbsp;    entereval\, but the corresponding JUMPENV\_PUSH\(3\) unwound all the way
to the
&nbsp;    outer perl\_run\, losing all the mg\_get\(\) related parts of the C stack\.

&nbsp;    It turns out that the run\-time parts of pp\_entereval were protected with
&nbsp;    a new JUMPENV level\, but the compile\-time parts weren't\. Add this\.

:100644 100644 80c7b221d7e967a6f3380d70917bd73700b16852
bbb2d1587cb197977c996ecfe8abddf4f9aa3631 M      pp\_ctl\.c
:040000 040000 7fa8480b6084b533c3eaddbbbd6834e2551f8a4e
f2ec8022695df42cff76ab78c117bd44d1fb626c M      t

So I'm closing this\. Tests for this particular issue in combination with
overload might be a good thing\.

**Comment by p5pRT** [comment]
@rafl - Status changed from 'open' to 'resolved'
